### PR TITLE
fix: correct calculation edge cases

### PR DESCRIPTION
**Pull request by hiroppy**
## Summary
- fix Japanese amount and percentage parsing for negative/unit edge cases
- avoid month-end rollover in history-mode month selection
- preserve zero unrealized gains and parse non-padded monthly summary dates
- save historical MM/DD transactions with the target month year
- fix analytics new-category detection and latest-vs-previous-3-month comparison

## Verification
- pnpm test
- pnpm typecheck
- pnpm format:check
- pnpm lint

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved number and percentage parsing (negative indicators, full-width signs, decimals)
  * Corrected month calculations and YYYY‑MM formatting around timezone boundaries and month rollovers
  * Fixed portfolio parsing so zero values and placeholder empties are preserved/ignored appropriately
  * Ensure MM/DD transaction dates are saved with the intended year
  * Refined trend and new‑category detection logic for comparisons

* **New Features**
  * Exposed helpers for consistent month range and monthly-header parsing used by history and monthly-summary flows

* **Tests**
  * Added/expanded tests for month rollovers, parsing, monthly summaries, portfolio parsing, transactions, and analyzers

### CODE REVIEW

**Comment by hiroppy** [comment]
@codex review

File: packages/db/src/queries/holding.ts [hunk 1]
```diff
@@ -6,13 +6,17 @@ const INVESTMENT_CATEGORIES = ["株式(現物)", "投資信託"];
...
-export async function getLatestSnapshot(db: Db = getDb()) {
-  return await db
-    .select()
-    .from(schema.dailySnapshots)
-    .orderBy(desc(schema.dailySnapshots.id))
+export async function getLatestSnapshot(db: Db = getDb(), groupId?: string | null) {
+  let query = db.select().from(schema.dailySnapshots).$dynamic();
+
+  if (groupId) {
+    query = query.where(eq(schema.dailySnapshots.groupId, groupId));
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply packages/db/src/queries/holding.ts]: Fixed in e77e57d by reverting the group-scoped snapshot change. `saveGroupOnlyData` does not create holding snapshots, so holdings should continue to use the global latest snapshot and then filter holdings by the selected group accounts.

File: apps/crawler/src/scrapers/monthly-summary.ts [hunk 2]
```diff
@@ -31,9 +31,9 @@ export async function scrapeMonthlySummary(page: Page): Promise<MonthlySummaryIt
   // 月を解析（最初のセルは空なのでスキップ）
   const months: string[] = [];
   for (let i = 1; i < headers.length; i++) {
-    const match = headers[i].trim().match(/^(\d{4})\/(\d{2})\/\d{2}〜$/);
+    const match = headers[i].trim().match(/^(\d{4})\/(\d{1,2})\/\d{1,2}〜$/);
     if (match) {
-      months.push(`${match[1]}-${match[2]}`);
+      months.push(`${match[1]}-${match[2].padStart(2, "0")}`);
     }
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/monthly-summary.ts]: Fixed in aeb89a6: extracted `parseMonthlySummaryMonths` and added unit coverage for padded, non-padded, and invalid monthly headers.

File: apps/crawler/src/scrapers/portfolio.ts [hunk 3]
```diff
@@ -150,7 +150,7 @@ async function parseStocks(page: Page): Promise<PortfolioItem[]> {
       avgCostPrice: orUndefined(parseDecimalNumber(avgCostText)),
       unitPrice: orUndefined(parseDecimalNumber(unitPriceText)),
       dailyChange,
-      unrealizedGain: parseJapaneseNumber(unrealizedGainText) || undefined,
+      unrealizedGain: unrealizedGainText ? parseJapaneseNumber(unrealizedGainText) : undefined,
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/portfolio.ts]: Fixed in aeb89a6: added `parseOptionalJapaneseNumber` coverage for `0`, empty text, and placeholder values. Also treats placeholders as `undefined` so they are not converted to 0.

**Comment by hiroppy** [comment]
@codex review

File: apps/crawler/src/history-months.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/**
+ * History mode fetches months by calendar month, not by day offset.
+ * Using Date#setMonth on month-end dates can roll over to the wrong month
+ * (e.g. 2026-03-31 minus 1 month becomes 2026-03-03).
+ */
+export function getHistoryMonth(now: Date, monthsAgo: number): string {
+  const date = new Date(now.getFullYear(), now.getMonth() - monthsAgo, 1);
```
Comment (hiroppy) [inline apps/crawler/src/history-months.ts]: ⚠️ **修正推奨:** ここは実行環境のローカルTZ依存になります。MoneyForward/ワークフローはJST基準ですが、`daily-update` は `ubuntu-latest` 上で動くため `Date#getFullYear()` / `getMonth()` はUTC基準です。たとえば JST `2026-07-01 06:50`（cron の朝実行）は UTC では `2026-06-30 21:50` なので、`monthsAgo=0` が `2026-06` になり、当月ではなく前月を history 対象にします。

JSTで年月を取り出す（`Intl.DateTimeFormat('ja-JP', { timeZone: 'Asia/Tokyo', ... })` など）か、呼び出し側からJST基準の年月を渡す形にした方が安全です。

**Review by hiroppy — COMMENTED** [review]
## Hermes Agent Review

GitHub制約で `COMMENT` として投稿しますが、内容としては修正推奨です。

確認したこと:
- 差分と周辺コード確認
- `pnpm test` / `pnpm typecheck` / `pnpm format:check` / `pnpm lint` 実行（すべて exit 0、lint は既存 warning のみ）
- GitHub Checks も pass

主に2点、月境界と新カテゴリ分析の edge case を inline でコメントしました。

File: packages/analytics/src/insights/analyze-spending-comparison.ts [hunk 5]
```diff
@@ -58,8 +58,9 @@ export function analyzeSpendingComparison(
   for (const [category, monthData] of byCategory) {
     const currentAmount = monthData.get(latestMonth) ?? 0;
     const prevMonthsForCat = allMonths.filter((m) => m < latestMonth);
+    const hasPreviousAmount = prevMonthsForCat.some((m) => (monthData.get(m) ?? 0) > 0);
```
Comment (hiroppy) [inline packages/analytics/src/insights/analyze-spending-comparison.ts:61]: ⚠️ **修正推奨:** 最新月で初出のカテゴリがあるケースで、`prev3Values` が `[0, 0, 0]` になり `threeMonthAvg=0` / `deviationFromThreeMonthPct=null` になります。そのカテゴリは `deviationFromThreeMonth > 0` なので `topIncreasing` に入り、戻り値型では `number` のはずの `diffPct` に `null` が返ります。

再現例: Jan〜Mar は食費のみ、Apr に医療費が初出 → `topIncreasing=[{ category: '医療費', diff: 10000, diffPct: null }]`。新カテゴリは `topIncreasing` から除外するか、`diffPct` の型/値を明示的に扱うのが良さそうです。

File: apps/crawler/src/history-months.ts (hunk 4 above)
Comment (hiroppy) [reply apps/crawler/src/history-months.ts]: Fixed in 70c8c3a: `getHistoryMonth` now extracts year/month in `Asia/Tokyo` via `Intl.DateTimeFormat` before doing UTC month arithmetic, so CI/UTC execution still targets the JST month. Added a regression test for JST 2026-07-01 06:50 (UTC 2026-06-30 21:50).

File: packages/analytics/src/insights/analyze-spending-comparison.ts (hunk 5 above)
Comment (hiroppy) [reply packages/analytics/src/insights/analyze-spending-comparison.ts:61]: Fixed in 70c8c3a: categories with a zero previous 3-month average now stay out of topIncreasing/topDecreasing unless a percentage diff exists, so `diffPct` is never null in those arrays. Added a regression test for a latest-month new category.

File: apps/crawler/src/index.ts [hunk 6]
```diff
@@ -147,9 +148,7 @@ async function main() {
       // DBにない最古の月まで取得
       let monthsToFetch = 1; // 今月は常に取得
       for (let i = 1; i < maxMonths; i++) {
-        const date = new Date();
-        date.setMonth(date.getMonth() - i);
-        const month = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}`;
+        const month = getHistoryMonth(now, i);
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/index.ts:150]: Fixed in 5854a87: `maxMonths` now uses `getHistoryMaxMonths(now)`, which derives the current month in `Asia/Tokyo` via the same helper as `getHistoryMonth`. Added a regression test for 2026-02-01 01:00 JST (UTC 2026-01-31 16:00), expecting 14 months so 2025-01 is included.

File: apps/crawler/src/scrapers/portfolio.ts [hunk 7]
```diff
@@ -103,6 +103,14 @@ function orUndefined(value: number): number | undefined {
   return value || undefined;
 }
 
+export function parseOptionalJapaneseNumber(text: string): number | undefined {
+  const trimmed = text.trim();
+  if (!trimmed || ["-", "—", "–", "―"].includes(trimmed)) {
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/scrapers/portfolio.ts]: Fixed in 5854a87: `parseOptionalJapaneseNumber` now treats the U+2212 minus glyph (`−`) as a missing optional value, matching the existing placeholder dash handling. Added unit coverage asserting `−` returns `undefined`.

**Comment by hiroppy** [comment]
@codex review
